11_rtc_driver/ds3231x_drv.c

```diff
@@ -0,0 +1,194 @@
+	struct ds3231x	*ds3231 = dev_get_drvdata(dev);
+	int		ret;
+	u8		regs[7];
+
+	dev_info(dev, "%s, time to write: secs=%d, mins=%d, hours=%d, mday=%d, mon=%d, year=%d, wday=%d\n",
+		__func__, time->tm_sec, time->tm_min,
+		time->tm_hour, time->tm_mday,
+		time->tm_mon, time->tm_year, time->tm_wday);
+
+
+	regs[DS3231_REG_SECS] = bin2bcd(time->tm_sec);
+	regs[DS3231_REG_MIN]  = bin2bcd(time->tm_min);
+	regs[DS3231_REG_HOUR] = bin2bcd(time->tm_hour);
+	regs[DS3231_REG_WDAY] = bin2bcd(time->tm_wday + 1);
+	regs[DS3231_REG_MDAY] = bin2bcd(time->tm_mday);
+	regs[DS3231_REG_MONTH] = bin2bcd(time->tm_mon + 1);
+	if (time->tm_year >= 100) {
+		regs[DS3231_REG_MONTH] |= DS3231_BIT_CENTURY;
+		regs[DS3231_REG_YEAR] = bin2bcd(time->tm_year - 100);
+	} else {
+		regs[DS3231_REG_YEAR] = bin2bcd(time->tm_year);
+	}
+
+	dev_info(dev, "%s, regs values: %7ph\n",  __func__, regs);
+
+	ret = regmap_bulk_write(ds3231->regmap, 0, regs,
+				   sizeof(regs));
+	if (ret) {
+		dev_err(dev, "%s, failed to write time regs: %d\n",
+			__func__, ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static const struct rtc_class_ops ds3231_rtc_ops = {
+	.read_time	= ds3231x_read_time,
+	.set_time	= ds3231x_set_time,
+};
+
+static int ds3231x_rtc_startup(struct device *dev)
+{
+	struct ds3231x *ds3231 = dev_get_drvdata(dev);
+	u8 ctl[2] = {0};
+	int ret = 0;
+
+	ret = regmap_bulk_read(ds3231->regmap, DS3231_REG_CONTROL, ctl, 2);
+	if (ret)
+		return ret;
+
+	if (ctl[1] & DS3231_BIT_OSF)
+		dev_warn(ds3231->dev, "oscillator was stopped, time value may be unreliable\n");
+
+	ctl[1] &= ~(DS3231_BIT_OSF | DS3231_BIT_A1I | DS3231_BIT_A2I);
+
+	ctl[0] &= ~(DS3231_BIT_A1IE | DS3231_BIT_A2IE);
+	ctl[0] |= DS3231_BIT_INTCN;
+
+	ret = regmap_bulk_write(ds3231->regmap, DS3231_REG_CONTROL, ctl, 2);
+
+	return ret;
+}
+
+static int ds3231x_probe(struct i2c_client *client,
+			 const struct i2c_device_id *id)
+{
+	static const struct regmap_config config = {
+		.reg_bits = 8,
+		.val_bits = 8,
+		.max_register = 0x13,
+	};
+
+	struct ds3231x *ds3231;
+	int ret;
+
+	ds3231 = devm_kzalloc(&client->dev, sizeof(*ds3231), GFP_KERNEL);
+	if (!ds3231)
+		return -ENOMEM;
+
+	ds3231->regmap = devm_regmap_init_i2c(client, &config);
+	if (IS_ERR(ds3231->regmap)) {
+		dev_err(&client->dev, "%s: failed to alloc regmap: %ld\n",
+			__func__, PTR_ERR(ds3231->regmap));
+		return PTR_ERR(ds3231->regmap);
+	}
+
+	ds3231->dev = &client->dev;
+
+	dev_set_drvdata(&client->dev, ds3231);
+
+	ret = ds3231x_rtc_startup(ds3231->dev);
+	if (ret) {
+		dev_err(ds3231->dev, "%s, failed to start up rtc\n", __func__);
+		return ret;
+	}
+
+	ds3231->rtc = devm_rtc_device_register(ds3231->dev, client->name,
+					       &ds3231_rtc_ops,
+						THIS_MODULE);
+	if (IS_ERR(ds3231->rtc)) {
+		dev_err(ds3231->dev, "%s, failed to redister rtc\n", __func__);
+		return PTR_ERR(ds3231->rtc);
+	}
+
+
+	return 0;
+}
+
+static struct i2c_driver ds3231x_driver = {
+	.driver = {
+		.name = "ds3231x",
+		.of_match_table = ds3231x_of_match,
+
+},
```

Comment:
indentation

Reply:
Fixed

Reply:
`},` should be moved one TAB to the right